src/ide/code_lens/mod.rs

```diff
@@ -18,17 +22,50 @@ pub struct CodeLensControllerState {
     lens: HashMap<Url, Vec<(CodeLens, CodeLensKind)>>,
 }
 
-#[derive(Clone, Default)]
+#[derive(Clone)]
 pub struct CodeLensController {
     state: Arc<RwLock<CodeLensControllerState>>,
+    refresh_sender: Sender<RefreshCodeLensRequest>,
+    request_refresh_receiver: Receiver<()>,
+    // Keep it last so we can drop channels.
+    // Otherwise, the refresh thread will never stop, and the it's
+    // JoinHandle drop will cause deadlock by waiting for the thread to join.
```

Comment:
```suggestion
    // Otherwise, the refresh thread will never stop, and the 
    // JoinHandle drop will cause deadlock by waiting for the thread to join.
```